Add our own versions of `FuturesUnordered` and `FuturesOrdered`

hrxi opened the pull request:
Unlike their `futures_util::stream::*` counterparts, they wake themselves when some future is pushed.

This gets rid of a couple of manual waker instances.

It also fixes some waking-related bugs (CC #2550).

Use clippy to warn for usages of the old types.

hrxi committed:
Add our own versions of `FuturesUnordered` and `FuturesOrdered`

Unlike their `futures_util::stream::*` counterparts, they wake themselves when
some future is pushed.

hrxi committed:
Use our own versions of `FuturesUnordered` and `FuturesOrdered`

This gets rid of a couple of manual waker instances.

It also fixes some waking-related bugs (CC #2550).

hrxi committed:
Warn for uses of `futures_util::stream::FuturesUnordered`

jsdanielh reviewed (commented):
Looks like we're still using the `futures`'s version in `consensus/src/sync/syncer.rs`.

hrxi committed:
Allow waking on non-empty peer list

This fixes a waking bug in the `Stream` implementation of `StateQueue`,
as it checked for a nonempty peer list but did not register a waker for
it.

CC #2550

Eligioo commented on consensus/src/sync/live/state_queue/mod.rs:
```diff
@@ -173,6 +173,8 @@ pub struct StateQueue<N: Network> {
     /// The blockchain event stream.
     blockchain_rx: BoxStream<'static, BlockchainEvent>,
+
+    nonempty_peers: Option<BoxFuture<'static, ()>>,
```
Add rustdoc

hrxi replied:
Thanks, fixed.

Eligioo commented on consensus/src/sync/live/state_queue/mod.rs:
```diff
@@ -568,9 +571,24 @@ impl<N: Network> Stream for StateQueue<N> {
-        if !self.chunk_request_component.has_pending_requests() && self.num_peers() > 0 {
-            self.request_chunk();
+        if self.nonempty_peers.is_none() && !self.chunk_request_component.has_pending_requests() {
+            println!("{}", self.request_chunk());
```
Forgot to remove the `println!`?

hrxi replied:
Thanks, fixed.

Eligioo commented on consensus/src/sync/peer_list.rs:
```diff
@@ -58,22 +60,27 @@ impl<N: Network> Default for PeerList<N> {
 impl<N: Network> Clone for PeerList<N> {
-    fn clone(&self) -> Self {
-        Self {
+    fn clone(&self) -> PeerList<N> {
```
Why can't both lines be `Self`?

hrxi replied:
Can both be `Self`. I deleted the implementation and recreated it during the creation of the PR, so I used the more explicit style. Interesting that it still ended up having such a small diff. I can change it to `Self` if you want.

Eligioo replied:
It's nitpicking but by having `Self` it can stay `Self`, even if the name of the struct changes :)

hrxi replied:
Changed it.

Eligioo reviewed (commented):
LGTM, some minor feedback points.

CC #2832 to use this new abstraction once this PR is in.

jsdanielh commented on utils/src/stream.rs:
```diff
@@ -0,0 +1,114 @@
+use crate::WakerExt as _;
+
+#[pin_project]
+pub struct FuturesOrdered<F: Future> {
```
I'd add some `rustdoc`. Something along these lines:
```
/// Wrapper over a `FuturesOredered` that auto-wakes on push
```

hrxi replied:
Done.

jsdanielh commented on utils/src/stream.rs:
```diff
@@ -0,0 +1,114 @@
+}
+
+#[pin_project]
+pub struct FuturesUnordered<F: Future> {
```
Same as above. I'd add some `rustdoc`

hrxi replied:
Done.

jsdanielh commented on utils/src/lib.rs:
```diff
@@ -18,6 +18,8 @@ pub mod tagged_signing;
 #[cfg(feature = "time")]
 pub mod time;
 
+pub mod stream;
```
Shouldn't we add a feature such that we don't pull unnecessary dependencies when we depend on this crate and we don't need anything on the stream module?

hrxi replied:
Done, although I doubt its usefulness. All of our crates depend on `futures` directly or indirectly.

jsdanielh commented on consensus/src/sync/live/state_queue/chunk_request_component.rs:
```diff
@@ -72,6 +73,11 @@ impl<N: Network> ChunkRequestComponent<N> {
         network.request::<RequestChunk>(request, peer_id).await
     }
+
+    pub fn notify_nonempty_peers(&self) -> Option<Arc<Notify>> {
```
Could you add some `rustdoc`?

hrxi replied:
Done.

jsdanielh commented on consensus/src/sync/live/state_queue/mod.rs:
```diff
@@ -568,8 +577,23 @@ impl<N: Network> Stream for StateQueue<N> {
         }
 
+        // Check if we have peers.
+        if self.nonempty_peers.is_none() {
```
Isn't it better to use `get_or_insert`?

hrxi replied:
This only inserts something if `self.chunk_request_component.notify_nonempty_peers()` returns a non-`None` value. This doesn't fit `Option::get_or_insert`'s function signature.

nibhar replied:
`notify_nonempty_peers()` can return `None` thus the following assignment can be to `None` again, which `get_or_insert` does not allow for.

jsdanielh commented on consensus/src/sync/live/state_queue/mod.rs:
```diff
@@ -568,8 +577,23 @@ impl<N: Network> Stream for StateQueue<N> {
+                .map(|notify| Box::pin(async move { notify.notified().await }) as BoxFuture<()>);
+        }
+        if let Some(nonempty_peers) = &mut self.nonempty_peers {
+            if nonempty_peers.as_mut().poll(cx).is_ready() {
```
Can't we just `take` it?

hrxi replied:
We only want to remove the future from the member variable when it resolves. `Option::take` can't do that.

jsdanielh commented on consensus/src/sync/peer_list.rs:
```diff
@@ -130,6 +137,10 @@ impl<N: Network> PeerList<N> {
         Some(self.peers[peer_index.index])
     }
+
+    pub fn notify_nonempty(&self) -> Arc<Notify> {
```
Could you add some `rustdoc`?

hrxi replied:
Done.

nibhar commented on consensus/src/sync/live/state_queue/mod.rs:
```diff
@@ -173,6 +173,14 @@ pub struct StateQueue<N: Network> {
+    /// `ChunkRequestComponent` when its peer list is nonempty, we need some
+    /// notification mechanism to wake us up once the list becomes nonempty if
+    /// we find it empty.
+    nonempty_peers: Option<BoxFuture<'static, ()>>,
```
I would prefer this to be named `peers_became_nonempty` or similar. This field does not hold any peers itself, but the name implies it to some extend.

hrxi replied:
Done.

nibhar commented on consensus/src/sync/peer_list.rs:
```diff
@@ -67,13 +70,17 @@ impl<N: Network> Clone for PeerList<N> {
 impl<N: Network> PeerList<N> {
     pub fn add_peer(&mut self, peer_id: N::PeerId) -> bool {
         if self.peers_set.insert(peer_id) {
+            if self.peers.is_empty() {
```
This seems weird, why does the set need to be empty to notify nonempty? Is there a guarantee that you cannot register on this Notify when the peer list is already non empty? 

Would there be any harm in just moving this down one line and inverting the condition?

That way whenever the set gets added to it will trigger a notify_waiters, which there should not be any if it was previously not empty, but if there are they now know it is (still) not empty. Otherwise they will be waiting forever as long as the list never gets empty again.

hrxi replied:
> This seems weird, why does the set need to be empty to notify nonempty?

The set only becomes nonempty when it was previously empty.

> Is there a guarantee that you cannot register on this Notify when the peer list is already non empty?

Not directly here, although I guess I could add it. The `notify_nonempty_peers` ensures it by returning an `Option` that only contains a notifier when the set is empty. I guess I should move that guarantee to this function.

> Would there be any harm in just moving this down one line and inverting the condition?
> 
> That way whenever the set gets added to it will trigger a notify_waiters, which there should not be any if it was previously not empty, but if there are they now know it is (still) not empty. Otherwise they will be waiting forever as long as the list never gets empty again.

The harm would mostly be hiding bugs and unnecessary wakeups. We could also wake this from many other places, so it's not clear that "adding a peer to an otherwise nonempty list" should trigger the "peer list became nonempty" any more than, say, "removing a peer from a list, and it's still nonempty".

hrxi replied:
> Is there a guarantee that you cannot register on this Notify when the peer list is already non empty?

I added that guarantee to the API now. Thanks for bringing it up.

nibhar commented on consensus/src/sync/peer_list.rs:
```diff
@@ -9,6 +10,7 @@ use nimiq_network_interface::network::Network;
 pub struct PeerList<N: Network> {
     peers_set: HashSet<N::PeerId>,
     peers: Vec<N::PeerId>,
+    notify_nonempty: Arc<Notify>,
```
This could probably do with some comment. I am also not too happy about the name, but right now have no better suggestion.

hrxi replied:
Added a comment.

nibhar commented on consensus/src/sync/sync_queue.rs:
```diff
@@ -22,7 +20,7 @@ use crate::sync::peer_list::PeerListIndex;
 #[pin_project]
 #[derive(Debug)]
-struct OrderWrapper<TId, TOutput> {
+struct OrderWrapper<TId: Clone, TOutput> {
```
Why does this need to be `Clone`? I cannot find it as new requirement.

hrxi replied:
True, removed it again.

hrxi committed:
nimiq-utils: Remove unused feature combinations

hrxi committed:
nimiq-utils: Introduce feature for `stream` module

nibhar commented on consensus/src/sync/sync_queue.rs:
```diff
@@ -452,14 +450,24 @@ where
                     return if self.ids_to_request.is_empty() || self.peers.read().is_empty() {
-                        Poll::Ready(None)
+                        self.waker.store_waker(cx);
+                        Poll::Pending
```
The return could be moved out of the conditional as they are now the same. As half the condition is repeated in the else case this can maybe also be combined.

hrxi replied:
Did the former. Not sure how to do the latter.

jsdanielh commented:
Rebasing branch to merge it
